impl/src/main/java/org/jboss/weld/xml/BeansXmlValidator.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2018, Red Hat, Inc., and individual contributors
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.weld.xml;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import javax.xml.XMLConstants;
+import javax.xml.transform.Source;
+import javax.xml.transform.stream.StreamSource;
+import javax.xml.validation.Schema;
+import javax.xml.validation.SchemaFactory;
+import javax.xml.validation.Validator;
+
+import org.jboss.weld.logging.XmlLogger;
+import org.xml.sax.ErrorHandler;
+import org.xml.sax.SAXException;
+import org.xml.sax.SAXParseException;
+
+/**
+ * This validator is thread safe and suitable for sharing between threads.
+ *
+ * @author Martin Kouba
+ */
+public class BeansXmlValidator implements ErrorHandler {
+
+    private static final Source[] EMPTY_SOURCE_ARRAY = new Source[0];
+    private static final String ROOT_ELEMENT_NAME = "beans";
+    // See also http://www.w3.org/TR/2001/REC-xmlschema-1-20010502/#cvc-elt
+    private static final String VALIDATION_ERROR_CODE_CVC_ELT_1 = "cvc-elt.1";
+
+    private final Schema cdi11Schema;
+
+    private final Schema cdi20Schema;
+
+    public BeansXmlValidator() {
+        SchemaFactory factory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+        cdi11Schema = initSchema(factory, XmlSchema.CDI11_SCHEMAS);
+        cdi20Schema = initSchema(factory, XmlSchema.CDI20_SCHEMAS);
+    }
+
+    public void validate(URL beansXml, ErrorHandler errorHandler) {
+        if (beansXml == null) {
+            throw XmlLogger.LOG.loadError("unknown", null);
+        }
+        if (errorHandler == null) {
+            errorHandler = this;
+        }
+        Schema schema = cdi20Schema;
```

Comment:
(Just a quick check: if there's no schema specified in the `beans.xml`, then CDI 2.0 will be used.  I assume that's right?)

Reply:
Yep.